### PR TITLE
ramda: add support of applying `ap` to function type argument

### PR DESCRIPTION
**Pull request by jituanlin**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://gist.github.com/jituanlin/168c2eb147f69b60cc063c08964cbf58>>
- [ ] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by googol** [comment]
Looks good. Adding a test, for example the one in the documentation would be a good addition :+1:

**Comment by DanielRosenwasser** [comment]
Can you add a test please?

**Comment by jituanlin** [comment]
> Can you add a test please?

OK, I will finish the test lately.

File: types/ramda/ramda-tests.ts [hunk 1]
```diff
@@ -2197,6 +2197,7 @@ class Rectangle {
 () => {
     const x: number[] = R.ap([R.multiply(2), R.add(3)], [1, 2, 3]); // => [2, 4, 6, 4, 5, 6]
     const y: number[] = R.ap([R.multiply(2), R.add(3)])([1, 2, 3]); // => [2, 4, 6, 4, 5, 6]
+    const z: string = R.ap<string, string, string>(R.concat, R.toUpper)('Ramda'); // => 'RamdaRAMDA'
```
Comment (DanielRosenwasser) [inline types/ramda/ramda-tests.ts]: Will this actually require providing type arguments? If not, please remove them. If so, then maybe we should consider the declaration here.

Comment (jituanlin) [reply types/ramda/ramda-tests.ts]: Provide the type arguments explicitly is required here.
This is cause by Typescript `strange` type infer behavior of overload function.
In this case, the `concat`  function not be infer `correctly`(you can see it be infer a very not make sense if you check the code in your editor), so we need provider type arguments
explicitly.

See [this issues](https://github.com/microsoft/TypeScript/issues/32081)

**Pull request merged**